File: src/main/java/yan/lx/bedrockminer/utils/CheckingEnvironment.java
package yan.lx.bedrockminer.utils;

import net.minecraft.block.Blocks;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.world.ClientWorld;
import net.minecraft.item.ItemPlacementContext;
import net.minecraft.util.Hand;
import net.minecraft.util.hit.BlockHitResult;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;

import java.util.ArrayList;

import static net.minecraft.block.Block.sideCoversSmallSquare;

public class CheckingEnvironment {

    public static BlockPos findNearbyFlatBlockToPlaceRedstoneTorch(ClientWorld world, BlockPos blockPos) {
        for (Direction direction : Direction.Type.HORIZONTAL) {
            Direction redstoneTorchFacing = Direction.UP;           // 红石火把的朝向
            BlockPos baseBlockPos = blockPos.offset(direction);     // 底座位置
            BlockPos redstoneTorchPos = baseBlockPos.offset(redstoneTorchFacing);   // 红石火把位置
            // 红石火把可以被放置情况下
            if (sideCoversSmallSquare(world, baseBlockPos, redstoneTorchFacing) && world.getBlockState(redstoneTorchPos).isReplaceable()) {
                return baseBlockPos;
            } else if (world.getBlockState(redstoneTorchPos).isOf(Blocks.REDSTONE_TORCH)) { // 红石火把已放置情况下
                return baseBlockPos;
            }
        }
        return null;
    }

    public static BlockPos findPossibleSlimeBlockPos(ClientWorld world, BlockPos blockPos) {
        for (Direction direction : Direction.Type.HORIZONTAL) {
            BlockPos newBlockPos = blockPos.offset(direction);
            // 粘液块无法被替换情况下
            if (!world.getBlockState(newBlockPos).isReplaceable()) {
                continue;
            }
            // 粘液块放置是否会被阻止
            if (CheckingEnvironment.isBlocked(newBlockPos)) {
                continue;
            }
            return newBlockPos;
        }
        return null;
    }

    public static boolean has2BlocksOfPlaceToPlacePiston(ClientWorld world, BlockPos blockPos) {
        if (world.getBlockState(blockPos.up()).getHardness(world, blockPos.up()) == 0) {
            BlockBreaker.breakBlock(world, blockPos.up());
        }
        if (isBlocked(blockPos.up())) {
            return false;
        }
        return world.getBlockState(blockPos.up().up()).isReplaceable();
    }

    public static ArrayList<BlockPos> findNearbyRedstoneTorch(ClientWorld world, BlockPos pistonBlockPos) {
        ArrayList<BlockPos> list = new ArrayList<>();
        if (world.getBlockState(pistonBlockPos.east()).isOf(Blocks.REDSTONE_TORCH)) {
            list.add(pistonBlockPos.east());
        }
        if (world.getBlockState(pistonBlockPos.west()).isOf(Blocks.REDSTONE_TORCH)) {
            list.add(pistonBlockPos.west());
        }
        if (world.getBlockState(pistonBlockPos.south()).isOf(Blocks.REDSTONE_TORCH)) {
            list.add(pistonBlockPos.south());
        }
        if (world.getBlockState(pistonBlockPos.north()).isOf(Blocks.REDSTONE_TORCH)) {
            list.add(pistonBlockPos.north());
        }
        return list;
    }

    public static boolean isBlocked(BlockPos blockPos) {
        ItemPlacementContext context = new ItemPlacementContext(MinecraftClient.getInstance().player,
                Hand.MAIN_HAND,
                Blocks.SLIME_BLOCK.asItem().getDefaultStack(),
                new BlockHitResult(blockPos.toCenterPos(), Direction.UP, blockPos, false));
        return !Blocks.SLIME_BLOCK.asItem().useOnBlock(context).isAccepted();
    }
}

File: src/client/kotlin/com/github/lxyan2333/bedrockminer/client/breaking/CheckingEnvironment.kt
package com.github.lxyan2333.bedrockminer.client.breaking

import net.minecraft.client.Minecraft
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.world.InteractionHand
import net.minecraft.world.item.context.BlockPlaceContext
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.Blocks
import net.minecraft.world.phys.BlockHitResult
import net.minecraft.world.phys.Vec3

object CheckingEnvironment {

    fun findNearbyFlatBlockToPlaceRedstoneTorch(level: Level, blockPos: BlockPos): BlockPos? {
        for (direction in Direction.Plane.HORIZONTAL) {
            val base = blockPos.relative(direction)
            val above = base.above()
            if (Block.canSupportCenter(level, base, Direction.UP) &&
                (level.getBlockState(above).canBeReplaced() || level.getBlockState(above).`is`(Blocks.REDSTONE_TORCH))
            ) {
                return base
            }
        }
        return null
    }

    fun findPossibleSlimeBlockPos(level: Level, blockPos: BlockPos): BlockPos? {
        for (direction in Direction.Plane.HORIZONTAL) {
            val pos = blockPos.relative(direction)
            if (!level.getBlockState(pos).canBeReplaced()) continue
            if (isBlocked(pos)) continue
            return pos
        }
        return null
    }

    fun has2BlocksOfPlaceToPlacePiston(level: Level, blockPos: BlockPos): Boolean {
        val above = blockPos.above()
        if (level.getBlockState(above).getDestroySpeed(level, above) == 0f) {
            BlockBreaker.breakBlock(above)
        }
        if (isBlocked(above)) return false
        return level.getBlockState(above.above()).canBeReplaced()
    }

    fun findNearbyRedstoneTorch(level: Level, pistonPos: BlockPos, axis: Direction.Axis): List<BlockPos> {
        return Direction.entries
            .filter { it.axis != axis }
            .map { pistonPos.relative(it) }
            .filter { level.getBlockState(it).`is`(Blocks.REDSTONE_TORCH) }
    }

    private fun isBlocked(blockPos: BlockPos): Boolean {
        val client = Minecraft.getInstance()
        val player = client.player ?: return true
        val context = BlockPlaceContext(
            player,
            InteractionHand.MAIN_HAND,
            Blocks.SLIME_BLOCK.asItem().defaultInstance,
            BlockHitResult(Vec3.atCenterOf(blockPos), Direction.UP, blockPos, false)
        )
        return !Blocks.SLIME_BLOCK.asItem().useOn(context).consumesAction()
    }
}